emasamplerate.go

```diff
@@ -0,0 +1,300 @@
+	var logSum float64
+	for _, count := range e.movingAverage {
+		logSum += math.Log10(float64(count))
+	}
+	goalRatio := goalCount / logSum
+
+	// must go through the keys in a fixed order to prevent rounding from changing
+	// results
+	keys := make([]string, len(e.movingAverage))
+	var i int
+	for k := range e.movingAverage {
+		keys[i] = k
+		i++
+	}
+	sort.Strings(keys)
+
+	// goal number of events per key is goalRatio * key count, but never less than
+	// one. If a key falls below its goal, it gets a sample rate of 1 and the
+	// extra available events get passed on down the line.
+	newSavedSampleRates := make(map[string]int)
+	keysRemaining := len(e.movingAverage)
+	var extra float64
+	for _, key := range keys {
+		count := e.movingAverage[key]
+		// take the max of 1 or my log10 share of the total
+		goalForKey := math.Max(1, math.Log10(count)*goalRatio)
+		// take this key's share of the extra and pass the rest along
+		extraForKey := extra / float64(keysRemaining)
+		goalForKey += extraForKey
+		extra -= extraForKey
+		keysRemaining--
+		if count <= goalForKey {
+			// there are fewer samples than the allotted number for this key. set
+			// sample rate to 1 and redistribute the unused slots for future keys
+			newSavedSampleRates[key] = 1
+			extra += goalForKey - count
+		} else {
+			// there are more samples than the allotted number. Sample this key enough
+			// to knock it under the limit (aka round up)
+			rate := math.Ceil(count / goalForKey)
+			// if counts are <= 1 we can get values for goalForKey that are +Inf
+			// and subsequent division ends up with NaN. If that's the case,
+			// fall back to 1
+			if math.IsNaN(rate) {
+				newSavedSampleRates[key] = 1
+			} else {
+				newSavedSampleRates[key] = int(rate)
+			}
+			extra += goalForKey - (count / float64(newSavedSampleRates[key]))
+		}
+	}
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	e.savedSampleRates = newSavedSampleRates
+	e.haveData = true
+
+	if e.testSignalMapsDone != nil {
+		e.testSignalMapsDone <- struct{}{}
+	}
+}
+
+// GetSampleRate takes a key and returns the appropriate sample rate for that
+// key. Will never return zero.
+func (e *EMASampleRate) GetSampleRate(key string) int {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	// Enforce MaxKeys limit on the size of the map
+	if e.MaxKeys > 0 {
+		// If a key already exists, increment it. If not, but we're under the limit, store a new key
+		if _, found := e.currentCounts[key]; found || len(e.currentCounts) < e.MaxKeys {
+			e.currentCounts[key]++
+			e.currentBurstSum++
+		}
+	} else {
+		e.currentCounts[key]++
+		e.currentBurstSum++
+	}
+
+	// Enforce the burst threshold
+	if e.burstThreshold > 0 && e.currentBurstSum >= e.burstThreshold && e.intervalCount >= e.BurstDetectionDelay {
+		// reset the burst sum to prevent additional burst updates from occurring while updateMaps is running
+		e.currentBurstSum = 0
+		// send but don't block - consuming is blocked on updateMaps, which takes the same lock we're holding
```

Comment:
what does the "consuming is blocked on updateMaps" part of that mean? 
Updating the sample rate map is going to wait, but `GetSampleRate` will still return without blocking, won't it?

Reply:
The issue I was thinking of here was - (assuming multiple GetSampleRate calls waiting on the mutex)
- burstSignal sent
- signal received, updateMaps starts, waits on lock
- GetSampleRate returns, releases lock
- Next GetSampleRate runs (since lock is free, and no guarantee updateMaps will be the first to grab it), detects burst, sends signal
- `select` still blocked on original call to `updateMaps` - now we're in deadlock

This is unlikely in our use case since `GetSampleRate` is only called in a single thread, but may happen in other applications.

Reply:
ah. I would update the comment to say "don't block if someone else has already sent the burst signal but it has not yet been processed". It's not the lock that's blocking in this case but pushing the burst signal on to the channel, right (which as a side effect holds the lock)?

Reply:
Yup exactly - it's the channel